packages/compartment-mapper/src/import-lite.js

```diff
@@ -85,16 +155,56 @@ export const loadFromMap = async (readPowers, compartmentMap, options = {}) => {
         exitModuleImportHook: compartmentExitModuleImportHook,
       },
     );
-    const { compartment, pendingJobsPromise } = link(compartmentMap, {
-      makeImportHook,
-      parserForLanguage,
-      languageForExtension,
-      globals,
-      transforms,
-      moduleTransforms,
-      __shimTransforms__,
-      Compartment,
-    });
+
+    /** @type {ImportNowHookMaker | undefined} */
+    let makeImportNowHook;
+
+    /** @type {Compartment} */
+    let compartment;
+    /** @type {Promise<void>} */
+    let pendingJobsPromise;
+
+    if (behavior.type === 'SYNC') {
+      const { importNowHook: exitModuleImportNowHook, syncModuleTransforms } =
+        behavior.options;
+      makeImportNowHook = makeImportNowHookMaker(
+        /** @type {SyncReadPowers} */ (readPowers),
+        entryCompartmentName,
+        {
+          compartmentDescriptors: compartmentMap.compartments,
+          searchSuffixes,
+          exitModuleImportNowHook,
+        },
+      );
+      ({ compartment, pendingJobsPromise } = link(compartmentMap, {
+        makeImportHook,
+        makeImportNowHook,
+        parserForLanguage,
+        languageForExtension,
+        globals,
+        transforms,
+        syncModuleTransforms,
+        __shimTransforms__,
+        Compartment,
+      }));
+    } else {
+      // sync module transforms are allowed, because they are "compatible"
+      // with async module transforms (not vice-versa)
+      const moduleTransforms = /** @type {ModuleTransforms} */ ({
+        ...behavior.options.syncModuleTransforms,
+        ...behavior.options.moduleTransforms,
+      });
+      ({ compartment, pendingJobsPromise } = link(compartmentMap, {
+        makeImportHook,
+        parserForLanguage,
+        languageForExtension,
+        globals,
+        transforms,
+        moduleTransforms,
+        __shimTransforms__,
+        Compartment,
+      }));
+    }
```

Comment:
I need to think more about what it would take to make this work with a single call to `link`, deferring the problem of differentiating these cases to the `link` implementation. That might be by passing a `sync: true` option into `link`.

Maybe this would give us more options: SES currently accepts both `importHook` and `importNowHook` but only consults one or the other. We could change `Compartment#import` to consult `importNowHook` before `importHook`, proceeding to the later only if the former returns `undefined`. Maybe that would let us get more bang from `importNowHook`. Or perhaps we just need that trampoline.

Reply:
TL;DR: link was synchronous and didn't care about all the async work happening around much, how about we keep it that way.

I also have an intuition that link could work differently, but my idea for where to take it is the opposite - make link totally colorblind and have one implementation of link that is fed pre-populated constructs/functions to use. 

eg. mapParsers choice depending on isSync and passing corresponding transforms could be prepared earlier and passed in as one mapParsers implementation. 

In that case the if on behavior would be used to collect the right type of tools and we'd only get one link function. 

Link doesn't do anything asynchronous itself, so threading a trampoline through link would be an unnecessary detour - the pre-cooking of some utilities it uses would be the opportunity to trampoline some of the implementations if necessary but the link.js file shouldn't see any of it.

The only new behavior in link as of today that needs to understand `isSync` is the decision around entry module, but that could also be solved with dependency injection - `conditionalImportNowHook<VerbHere>er` would accept all ingredients for importNowHookMaker and decide whether to use it or provide one that throws an informative error. (could be added on to importNowHookMaker to skip using the new name I just half-made up)

I'm very self conscious about using lots of dependency injection vs just importing functionality as it requires jumping back and forth between files while preserving context to figure out which code actually runs, but here DI might actually concentrate context-dependent behavior more by removing that concern from link entirely and having a dense code block that makes all behavior decisions in one place. (note this is all trampoline-agnostic and the implementations of behaviors can be shared&bounced or not)

<details>
  <summary>alternatively</summary>
100% separate sync/async implementations exported from different modules sharing some of the internals and *C macros* in imports :see_no_evil: 
</details>